File: 2025/994. Rotting Oranges/Solution.java
class Solution {

    int[][] dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public int orangesRotting(int[][] grid) {
        int count = 0;
        int ones = 0;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    ones ++;
                } else if (grid[i][j] == 2) {
                    queue.add(new int[]{i, j});
                }
            }
        }
        while (!queue.isEmpty() && ones > 0) {
            int size = queue.size();
            count++;
            for (int i = 0; i < size; i++) {
                int[] point = queue.remove();
                for (int j = 0; j < 4; j++) {
                    int x = point[0] + dir[j][0];
                    int y = point[1] + dir[j][1];
                    if (x >= 0 && y >= 0 && x < grid.length && y < grid[0].length && grid[x][y] == 1) {
                        ones--;
                        grid[x][y] = 2;
                        queue.add(new int[]{x, y});
                    }
                }
            }
        }
        if (ones != 0) {
            return -1;
        }
        return count;
    }
}

File: 2025/994. Rotting Oranges/solution.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        queue = deque()
        fresh_count = rotten_count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 2:
                    queue.append((i,j))
                elif grid[i][j] == 1:
                    fresh_count += 1
        if fresh_count == 0:
            return 0
        while queue:
            next_set = set()
            oranges = []
            rotten_count += 1
            for _ in range(len(queue)):
                (i, j) = queue.popleft()
                for (x,y) in ((0,1),(0,-1),(1,0),(-1,0)):
                    ii, jj = i + x, j + y 
                    if (0 <= ii < len(grid) and 
                        0 <= jj < len(grid[0]) and
                        (ii, jj) not in next_set and
                        grid[ii][jj] == 1):
                        grid[ii][jj] = 2
                        fresh_count -= 1
                        queue.append((ii, jj))
        return rotten_count - 1 if not fresh_count else -1